spot-contracts/contracts/RolloverVault.sol

```diff
@@ -518,13 +491,162 @@ contract RolloverVault is
+
+        // We deduct fees by transferring out fewer perp tokens
+        perpAmtOut = perpAmtOut.mulDiv(FEE_ONE_PERC - (swapFeePerpSharePerc + swapFeeVaultSharePerc), FEE_ONE_PERC);
+
+        return (perpAmtOut, perpFeeAmtToBurn, s);
+    }
+
+    /// @notice Computes the amount of underlying tokens that are returned when user swaps a given number of perp tokens.
+    /// @param perpAmtIn The number of perp tokens the user swaps in.
+    /// @return underlyingAmtOut The number of underlying tokens returned to the user.
+    /// @return perpFeeAmtToBurn The amount of perp tokens to be paid to the perp contract as burn fees.
+    /// @return s The pre-swap perp and vault subscription state.
+    function computePerpToUnderlyingSwapAmt(uint256 perpAmtIn)
+        public
+        returns (
+            uint256,
+            uint256,
+            IFeePolicy.SubscriptionParams memory
+        )
+    {
+        // Compute equal value underlying tokens to swap out
+        IFeePolicy.SubscriptionParams memory s = _querySubscriptionState();
+        uint256 underlyingAmtOut = perpAmtIn.mulDiv(s.perpTVL, perp.totalSupply());
+
+        //-----------------------------------------------------------------------------
+        // When user swaps perps for vault's underlying -> perps are redeemed by the vault
+        // We thus compute fees based on the post-burn subscription state.
+        IFeePolicy.SubscriptionParams memory postSwapState;
+        postSwapState.perpTVL = s.perpTVL - underlyingAmtOut;
+        postSwapState.vaultTVL = s.vaultTVL;
+        postSwapState.perpTR = s.perpTR;
+        postSwapState.vaultTR = s.vaultTR;
+        (uint256 swapFeePerpSharePerc, uint256 swapFeeVaultSharePerc) = feePolicy.perpToUnderlyingSwapFeePercs(
+            feePolicy.computeSubscriptionRatio(postSwapState)
+        );
+        //-----------------------------------------------------------------------------
+
+        // Calculate perp fee share to be paid by the vault
+        uint256 perpFeeAmtToBurn = perpAmtIn.mulDiv(swapFeePerpSharePerc, FEE_ONE_PERC, MathUpgradeable.Rounding.Up);
+
+        // We deduct fees by transferring out fewer underlying tokens
+        underlyingAmtOut = underlyingAmtOut.mulDiv(
+            FEE_ONE_PERC - (swapFeePerpSharePerc + swapFeeVaultSharePerc),
+            FEE_ONE_PERC
+        );
+
+        return (underlyingAmtOut, perpFeeAmtToBurn, s);
+    }
+
+    //--------------------------------------------------------------------------
+    // External & Public read methods
+
+    /// @inheritdoc IVault
+    function assetCount() external view override returns (uint256) {
+        return _deployed.length() + 1;
+    }
+
+    /// @inheritdoc IVault
+    function assetAt(uint256 i) external view override returns (IERC20Upgradeable) {
+        if (i == 0) {
+            return underlying;
+        } else if (i < _deployed.length() + 1) {
```

Comment:
I like the clarity how it is, but slight gas improvement with
`i <= _deployed.length()`